cmd/configurator/main.go

```diff
@@ -0,0 +1,317 @@
+	if err != nil {
+		log.Fatalf("can't fetch cluster config %s", err)
+	}
+
+	// corev1 client set is required to create collecting results pod.
+	coreC, err := corev1.NewForConfig(config)
+	if err != nil {
+		log.Fatalf("can't create core client %s", err)
+	}
+
+	conn, err := grpc.Dial("unix://"+*redBoxSock, grpc.WithInsecure())
+	if err != nil {
+		log.Fatalf("can't connect to %s %s", *redBoxSock, err)
+	}
+	slurmC := api.NewWorkloadManagerClient(conn)
+
+	ctx, cancel := context.WithCancel(context.Background())
+
+	wg := &sync.WaitGroup{}
+	wg.Add(1)
+	go watchPartitions(ctx, wg, slurmC, coreC)
+
+	sig := make(chan os.Signal, 1)
+	signal.Notify(sig, unix.SIGINT, unix.SIGTERM, unix.SIGQUIT)
+
+	log.Printf("Got signal %s", <-sig)
+	cancel()
+
+	wg.Wait()
+
+	log.Println("Configurator is finished")
+}
+
+func watchPartitions(ctx context.Context, wg *sync.WaitGroup, slurmClient api.WorkloadManagerClient, k8sClient *corev1.CoreV1Client) {
+	defer wg.Done()
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		case <-time.Tick(1 * time.Minute):
+			// getting SLURM partitions
+			partitionsResp, err := slurmClient.Partitions(context.Background(), &api.PartitionsRequest{})
+			if err != nil {
+				log.Printf("Can't get partitions %s", err)
+				continue
+			}
+
+			// gettings k8s nodes
+			nodes, err := k8sClient.Nodes().List(metav1.ListOptions{
+				LabelSelector: "type=virtual-kubelet",
+			})
+			if err != nil {
+				log.Printf("Can't get pods %s", err)
+				continue
+			}
+			// extract partition names from k8s nodes
+			nNames := partitionNames(nodes.Items)
+
+			// check which partitions are not yet represented in k8s
+			partitionToCreate := notIn(partitionsResp.Partition, nNames)
+			// creating pods for that partitions
+			if err := createNodeForPartitions(k8sClient, partitionToCreate); err != nil {
+				log.Printf("Can't create partitions  %s", err)
+			}
+
+			// some partitions can be deleted from SLURM, so we need to delete pods
+			// which are represent those deleted partitions
+			nodesToDelete := notIn(nNames, partitionsResp.Partition)
+			if err := deleteControllingPod(k8sClient, nodesToDelete); err != nil {
+				log.Printf("Can't delete controlling pod %s", err)
+			}
+		}
+
+	}
+}
+
+func createNodeForPartitions(k8sClient *corev1.CoreV1Client, partitions []string) error {
+	for _, p := range partitions {
+		log.Printf("Creating pod for %s partition in %s namespace", p, namespace)
+		_, err := k8sClient.Pods(namespace).
+			Create(virtualKubeletPodTemplate(p, hostNodeName))
+		if err != nil {
+			return errors.Wrap(err, "can't create pod for partition")
+		}
+	}
+
+	return nil
+}
+
+func deleteControllingPod(k8sClient *corev1.CoreV1Client, nodes []string) error {
+	for _, n := range nodes {
+		nodeName := partitionNodeName(n, hostNodeName)
+		log.Printf("Deleting pod %s in %s namespace", nodeName, namespace)
+		if err := k8sClient.Pods(namespace).Delete(nodeName, &metav1.DeleteOptions{}); err != nil {
```

Comment:
can be simplified to errors.Wrap without if
b/c errors automatically check for nil